Applications/VpView/vpVdfTrackIO.cxx

```diff
@@ -0,0 +1,266 @@
+/*ckwg +5
+ * Copyright 2018 by Kitware, Inc. All Rights Reserved. Please refer to
+ * KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
+ * Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.
+ */
+
+#include "vpVdfTrackIO.h"
+
+#include "vpFrameMap.h"
+#include "vpVdfIO.h"
+#include "vtkVpTrackModel.h"
+
+#include <vdfDataSource.h>
+#include <vdfSourceService.h>
+#include <vdfTrackReader.h>
+
+#include <qtEnumerate.h>
+#include <qtStlUtil.h>
+
+#include <vtksys/Glob.hxx>
+
+#include <QUrl>
+
+namespace
+{
+
+//-----------------------------------------------------------------------------
+void addPoint(std::vector<float>& points, float x, float y, float z = 0.0f)
+{
+  points.push_back(x);
+  points.push_back(y);
+  points.push_back(z);
+}
+
+} // namespace <anonymous>
+
+QTE_IMPLEMENT_D_FUNC(vpVdfTrackIO)
+
+//-----------------------------------------------------------------------------
+class vpVdfTrackIOPrivate
+{
+public:
+  vpVdfIO* Base;
+  QUrl TracksUri;
+};
+
+//-----------------------------------------------------------------------------
+vpVdfTrackIO::vpVdfTrackIO(
+  vpVdfIO* base, vtkVpTrackModel* trackModel,
+  TrackStorageMode storageMode, TrackTimeStampMode timeStampMode,
+  vtkVgTrackTypeRegistry* trackTypes, vtkMatrix4x4* geoTransform,
+  vpFrameMap* frameMap)
+  : vpTrackIO{trackModel, storageMode, timeStampMode,
+              trackTypes, geoTransform, frameMap},
+    d_ptr{new vpVdfTrackIOPrivate}
+{
+  QTE_D();
+  d->Base = base;
+}
+
+//-----------------------------------------------------------------------------
+vpVdfTrackIO::~vpVdfTrackIO()
+{
+}
+
+//-----------------------------------------------------------------------------
+unsigned int vpVdfTrackIO::GetImageHeight() const
+{
+  QTE_D();
+  return d->Base->GetImageHeight();
+}
+
+//-----------------------------------------------------------------------------
+void vpVdfTrackIO::SetTracksUri(const QUrl& uri)
+{
+  QTE_D();
+  d->TracksUri = uri;
+}
+
+//-----------------------------------------------------------------------------
+bool vpVdfTrackIO::ReadTracks()
+{
+  QTE_D();
+
+  vdfTrackReader reader;
+
+  if (d->TracksUri.isLocalFile())
+    {
+    // Interpret the specifier as a glob
+    vtksys::Glob glob;
+    glob.FindFiles(stdString(d->TracksUri.toLocalFile()));
+    glob.SetRecurse(true);
+    const auto& files = glob.GetFiles();
+    if (files.empty())
+      {
+      return false;
+      }
+
+    // Read through each track file
+    for (size_t i = 0, k = files.size(); i < k; ++i)
+      {
+      // Construct the track source and track reader
+      const auto& trackUri = QUrl::fromLocalFile(qtString(files[i]));
+      QScopedPointer<vdfDataSource> source{
+        vdfSourceService::createArchiveSource(trackUri)};
+
+      if (source)
+        {
+        reader.setSource(source.data());
+
+        // Read tracks
+        if (!reader.exec() && reader.failed())
+          {
+          // Track reading failed; die
```

Comment:
I'm not sure what the use case is for multiple track files, but if somehow have multiple and fail to read one of the files, then don't load any data.

Would this code even work if there were multiple track files (that did each load successfully)?

Not saying we should do this differently - just verifying some thought on the matter and that it is acting as intended (which is maybe different than I was thinking)

Reply:
I have a vague recollection that @mattdawkins had a use case where users had one file per track, or some such thing... I think it should work, as long as the tracks that are read don't overlap (and/or if they do, "last one wins" will be in play). And, yes, this is set up as an all-or-nothing load.

Reply:
seems to be same behavior as elsewhere - just wanted to raise some awareness of the situation.